File: day7/main.py
from functools import lru_cache

input_file = "input.txt"


@lru_cache(None)  # Memoization to cache intermediate results
def bt(total, values, value_index, target):
    if total == target and value_index == len(values):
        return True
    if value_index >= len(values):  # Stop if all values are used and target not reached
        return False

    # Explore the next value with operations
    current_value = values[value_index]
    if bt(total * current_value, values, value_index + 1, target):
        return True
    if bt(total + current_value, values, value_index + 1, target):
        return True

    return False


@lru_cache(None)
def bt2(total, values, value_index, target):
    if total == target and value_index == len(values):
        return True
    if value_index >= len(values):  # Stop if all values are used and target not reached
        return False

    # Explore the next value with operations
    current_value = values[value_index]
    if bt2(total * current_value, values, value_index + 1, target):
        return True
    if bt2(total + current_value, values, value_index + 1, target):
        return True
    new_total = total * (10 ** len(str(current_value))) + current_value
    if bt2(new_total, values, value_index + 1, target):
        return True

    return False


# part 1
def p1():
    with open(input_file) as f:
        out = 0
        lines = f.read().splitlines()
        for line in lines:
            contents = line.split()
            target = int(contents[0][:-1])

            values = tuple(int(x) for x in contents[1:])  # Convert to tuple

            # Use backtracking to check if any are valid
            if bt(values[0], values, 1, target):
                out += target
        print(out)


# part 2
def p2():
    with open(input_file) as f:
        out = 0
        lines = f.read().splitlines()
        for line in lines:
            contents = line.split()
            target = int(contents[0][:-1])

            values = tuple(int(x) for x in contents[1:])  # Convert to tuple

            # Use backtracking to check if any are valid
            if bt2(values[0], values, 1, target):
                out += target
        print(out)


p1()
p2()

File: day7/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

fn part1() -> std::io::Result<i32> {
    let mut out = 0;
    let f = File::open("input.txt")?;
    let reader = BufReader::new(f);
    let lines = reader.lines();
    for line in lines {
        let line = line?;
    }

    Ok(out)
}
fn part2() -> std::io::Result<i32> {
    let mut out = 0;
    let f = File::open("input.txt")?;
    let reader = BufReader::new(f);
    let lines = reader.lines();
    for line in lines {
        let line = line?;
    }

    Ok(out)
}

fn main() {
    let p1_result = part1().unwrap();
    let p2_result = part2().unwrap();
    println!("Part 1: {}\nPart 2: {}", p1_result, p2_result);
}
